storage/pin/pin_test.go

```diff
@@ -0,0 +1,509 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pin
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"io/ioutil"
+	"mime"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/swarm/api"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/feed"
+	"github.com/ethersphere/swarm/storage/localstore"
+	"github.com/ethersphere/swarm/testutil"
+)
+
+// TestPinRawUpload pins a RAW file and unpin it multiple times
+func TestPinRawUpload(t *testing.T) {
+	p, f, closeFunc := getPinApiAndFileStore(t)
+	defer closeFunc()
+
+	data := testutil.RandomBytes(1, 10000)
+	hash := uploadFile(t, f, data, false)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 3, true)
+}
+
+// TestPinRawUploadEncrypted pins a encrypted RAW file and unpin it multiple times
+func TestPinRawUploadEncrypted(t *testing.T) {
+	p, f, closeFunc := getPinApiAndFileStore(t)
+	defer closeFunc()
+
+	data := testutil.RandomBytes(2, 10000)
+	hash := uploadFile(t, f, data, true)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 3, true)
+}
+
+// TestPinCollectionUpload pins a simple collection and unpin it multiple times
+func TestPinCollectionUpload(t *testing.T) {
+	p, f, closeFunc := getPinApiAndFileStore(t)
+	defer closeFunc()
+
+	hash := uploadCollection(t, p, f, false)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 3, false)
+}
+
+// TestPinCollectionUploadEncrypted pins a encrypted simple collection and unpin it multiple times
+func TestPinCollectionUploadEncrypted(t *testing.T) {
+	p, f, closeFunc := getPinApiAndFileStore(t)
+	defer closeFunc()
+
+	hash := uploadCollection(t, p, f, true)
+
+	// test pin and unpin
+	pinUnpinAndFailIfError(t, p, hex.EncodeToString(hash), 3, false)
+}
+
+// TestWalker tests the walkChunksFromRootHash function which is the crux of
+// commands like pin, unpin & list.
+func TestWalker(t *testing.T) {
```

Comment:
@jmozah thanks for writing this test, but it:
1. does not create any manifests
2. if they were to be created, the manifests would not span more than one chunk
i'd expect this test to create a directory structure with nested directories and multiple files in the directories (to force creating multiple manifest nodes), then check the chunks created and traversed by the walker.
also, the usage of `GetAllReferences` is incorrect here. you will have to use `SubscribePull` to get all of the chunks in the localstore after uploading

Reply:
If i create manifest.. then it is exactly as the collection test. 
The intention of this test was to test the Walker only... thats what you wanted right.. test walker.. 

If you test only walker.. GetAllRef is enough.. Subscribe pull is needed only if we use manifest and upload a directory (which is also in the collection testcases)

So.. as a summary

- If you want to upload a directory tree and check walker with subscribePull.. there is collection testcases
- if you want to test walker only... there is this testcase...

Both ways covered.. per me.